### ISSUE TITLE
sr_cpost watch attempts to post non-utf8 paths on recursive deletion of directory with contents

### ISSUE THREAD

**Issue by msaraga** [MetPX/sarrac#57]
it was found that deleting a directory containing a file with `rm -rf dir_name` resulted in the following output:
```
aspymis@amqpbot:~/cpost_test$ sr_cpost -c cpost_test.conf -sum 0
2019-05-23 13:01:49,380 [INFO] cpost 2.19.05b1ubuntu2 config: cpost_test, pid: 17553, starting
2019-05-23 13:02:11,881 [INFO] published: 20190523130211.880591284 file:/ /home/aspymis/cpost_test/dd/s/t topic=v02.post.home.aspymis.cpost_test.dd.s sum=0,958 source=bunnymaster to_clusters=localhost from_cluster=localhost mtime=20190523130211.876 atime=20190523130211.876 mode=0664 parts=1,0,1,0,0
2019-05-23 13:02:18,494 [INFO] published: 20190523130218.494300923 file:/ /home/aspymis/cpost_test/dd/s/t topic=v02.post.home.aspymis.cpost_test.dd.s sum=R,99f97d455d5d62b24f3a942a1abc3fa8863fc0ce2037f52f09bd785b22b800d4f2e7b2b614cb600ffc2a4fe24679845b24886d69bb776fcfa46e54d188889c6f source=bunnymaster to_clusters=localhost from_cluster=localhost
2019-05-23 13:02:18,495 [ERROR] file path "/home/aspymis/cpost_test/dd/s/" not utf8 encoded, ignoring sr_post call
2019-05-23 13:02:18,496 [INFO] published: 20190523130218.495899027 file:/ /home/aspymis/cpost_test/dd/s topic=v02.post.home.aspymis.cpost_test.dd sum=R,2c1ee68372215b1ce064426b5cdbd4ef2581ace0dd3b21fa2be27f364827242e83f68b68be03f5b3e24be5d1b4315f98a0a96d19713fb3a19dc455fb6adc3431 source=bunnymaster to_clusters=localhost from_cluster=localhost
```

cpost_test.conf:
```
post_broker amqp://bunnymaster@localhost/
post_exchange xtest

path /home/aspymis/cpost_test/dd

sleep 1

post_base_url file:/

events create,delete,link,modify

accept .*
```

Command log:
```
aspymis@amqpbot:~/cpost_test/dd$ mkdir s
aspymis@amqpbot:~/cpost_test/dd$ touch s/t
aspymis@amqpbot:~/cpost_test/dd$ rm -rf s
```

this behaviour was first noted in #36

**Comment by petersilva** [comment]
It's actually worse than that... it doesn't notice anything else ever again, like the thing we see on the radar machines... it stops noticing any files at all.  I did
```

mkdir b
echo hoho b/lala
echo hahah b/lala2

```
I have debug on in my configuration, and as soon as I see:
`
detected directory removal, removing from internal data structures
`
It''s game over. no more events will be noticed.

**Comment by petersilva** [comment]
also get core dumps sometimes.

**Comment by petersilva** [comment]
```
2019-06-12 21:30:10,533 [INFO] published: 20190613013010.533142005 sftp://sarra_test@localhost home/peter/src/sarrac/test_dir/lalala topic=v02.post.home.peter.src.sarrac.test_dir sum=d,a8d2fbfa3ebfb655c428ce111202b2c8 source=tsource to_clusters=localhost from_cluster=localhost mtime=20190613013010.529298558 atime=20190613013010.529298558 mode=0644 parts=1,5,1,0,0 toto=pig
2019-06-12 21:30:41,617 [DEBUG] bytes read: 32, sz ev: 32, event: 40000200 delete,directory: len=16, fn=/home/peter/src/sarrac/test_dir/c
2019-06-12 21:30:41,617 [DEBUG] detected directory removal, removing from internal data structures
Program received signal SIGSEGV, Segmentation fault.
0x0000555555555be5 in dir_stack_rm (fn=0x7fffffff79c0 "/home/peter/src/sarrac/test_dir/c") at sr_cpost.c:155
155			i->next = i->next->next;
(gdb) 
(gdb) 
(gdb) 
```

**Comment by petersilva** [comment]
Added a fix to have it not crash at that line, then I see this:
`
 bytes read: 112, sz ev: 16, event: 8000 dunno: 8000!: len=0, fn=/home/peter/src/sarrac/test_dir/`

then found the .h file (/usr/include/sys/inotify.h, to determing:
`
#define IN_IGNORED		0x00008000	/* File was ignored */
`
then found this in the inotify(7) man page:
```

           IN_DELETE_SELF
                  Watched file/directory was itself deleted.  (This event
                  also occurs if an object is moved to another filesystem,
                  since mv(1) in effect copies the file to the other
                  filesystem and then deletes it from the original filesys‐
                  tem.)  In addition, an IN_IGNORED event will subsequently
                  be generated for the watch descriptor.
later...
      The following bits may be set in the mask field returned by read(2):

           IN_IGNORED
                  Watch was removed explicitly (inotify_rm_watch(2)) or
                  automatically (file was deleted, or filesystem was
                  unmounted).  See also BUGS.

           IN_UNMOUNT
                  Filesystem containing watched object was unmounted.  In
                  addition, an IN_IGNORED event will subsequently be gener‐
                  ated for the watch descriptor.
    
later...

       Suppose an application is watching the directory dir and (the empty)
       directory dir/subdir.  The following examples show some events that
       may be generated.

           mkdir("dir/new", mode);
                  Generates an IN_CREATE | IN_ISDIR event for dir.

           rmdir("dir/subdir");
                  Generates IN_DELETE_SELF and IN_IGNORED events for subdir,
                  and an IN_DELETE | IN_ISDIR event for dir.

```
I don't know if the IN_IGNORED is happenning to tell us that since the directory was deleted, that broke the inotify descriptor, and since we are using a single descriptor for the entire tree, we now have to do a priming read again?   that looks like what is happenning because subsequent sub-directory and file creations don't seem to be triggering events...

We should probably try ignoring the event first.

**Comment by petersilva** [comment]
also the code doesn't detect mkdir... in the trees at all.  I'll make a separate bug for that.

**Comment by petersilva** [comment]
no the descriptor is still fine for existing directories... just mkdirs don't get noticed. we have #59 for that.

**Comment by petersilva** [comment]
This is fixed by baca797a3a584d780e4db18f5c9900f9494f2266
